peps/pep-0705.rst

```diff
@@ -185,111 +381,180 @@ Discussion:
+    class ExampleUpdateDict(TypedDict, total=False, readonly=True, other_keys=False):
+        a: int
+        b: str
+        # c is not present as it is read-only in Example
 
+    ExampleUpdateEntry = tuple[Literal["a"], int] | tuple[Literal["b"], str]
+    ExampleUpdate = ExampleUpdateDict | Iterable[ExampleUpdateEntry]
 
-Security Implications
+Type checkers should permit any type compatible with this TypedDict to be passed into the update operation. As with :pep:`589`, they may choose to continue permitting TypedDict types that allow other keys as well, to avoid generating false positives.
+
+Backwards compatibility
+=======================
+
+This PEP changes the rules for how TypedDict behaves, so code that inspects TypedDict types will have to change to support types using the new features. This is expected to mainly affect type-checkers.
+
+Security implications
 =====================
 
 There are no known security consequences arising from this PEP.
 
-
 How to Teach This
 =================
 
-Class documentation should be added to the :mod:`typing` module's documentation, using
-that for :class:`~collections.abc.Mapping`, :class:`~typing.Protocol` and
-:class:`~typing.TypedDict` as examples. Suggested introductory sentence: "Base class
-for read-only mapping protocol classes."
+Suggestion for changes to the :mod:`typing` module, in line with current practice:
+
+* Add this PEP to the others listed.
+* Add ``typing.ReadOnly``, linked to TypedDict and this PEP.
+* Add the following text to the TypedDict entry:
 
-This PEP could be added to the others listed in the :mod:`typing` module's documentation.
+By default, keys not specified in a TypedDict may still be present. Instances can be restricted to only the named keys with the ``other_keys`` flag. *insert example, perhaps using ``in`` to illustrate the benefit*
 
+Individual keys can be excluded from mutate operations using ReadOnly, allowing them to be read but not changed. This is useful when the exact type of the value is not known yet, and so modifying it would break structural subtypes. *insert example*
+
+If all keys on a TypedDict should be read-only, the ``readonly`` flag can be used as a shorthand. *insert example*
 
 Reference Implementation
 ========================
 
-No reference implementation exists yet.
-
+No complete reference implementation exists yet. pyright 1.1.310 ships with a partial implementation of the ReadOnly qualifier.
 
 Rejected Alternatives
 =====================
 
-Several variations were considered and discarded:
+A TypedMapping protocol type
+----------------------------
+
+:pep:`705` proposed a ``TypedMapping`` protocol type, behaving much like a read-only TypedDict but without the constraint that the runtime type be a ``dict``. The behavior described in this PEP could then be obtained by inheriting a TypedDict from a TypedMapping. This has been set aside for now as more complex, without a strong use-case motivating the additional complexity.
```

Comment:
```suggestion
An earlier version of :pep:`705` proposed a ``TypedMapping`` protocol type, behaving much like a read-only TypedDict but without the constraint that the runtime type be a ``dict``. The behavior described in the current version of this PEP could then be obtained by inheriting a TypedDict from a TypedMapping. This has been set aside for now as more complex, without a strong use-case motivating the additional complexity.
```

This is PEP 705; the reference will be confusing to future readers

Reply:
Hahaha thanks, good catch! I started out by creating a new PEP but was advised to merge back into this one. Missed this bit of text!